server/src/test/java/org/elasticsearch/index/mapper/ObjectMapperMergeTests.java

```diff
@@ -109,29 +109,119 @@ public void testMergeDisabledRootMapper() {
+            "_doc",
+            true,
+            Collections.singletonMap("foo", createObjectSubobjectsFalseLeafWithDots())
+        );
+        RootObjectMapper mergeWith = createRootObjectMapper(
+            "_doc",
+            true,
+            Collections.singletonMap("foo", createObjectSubobjectsFalseLeafWithDots())
+        );
+
+        final ObjectMapper merged = mergeInto.merge(mergeWith, MapperBuilderContext.ROOT);
+
+        ObjectMapper foo = (ObjectMapper) merged.getMapper("foo");
+        ObjectMapper metrics = (ObjectMapper) foo.getMapper("metrics");
+        final KeywordFieldMapper keywordFieldMapper = (KeywordFieldMapper) metrics.getMapper("host.name");
+        assertEquals("foo.metrics.host.name", keywordFieldMapper.name());
+        assertEquals("host.name", keywordFieldMapper.simpleName());
+    }
+
+    public void testMergeMultiFields() {
+        RootObjectMapper mergeInto = createRootObjectMapper(
+            "_doc",
+            true,
+            Collections.singletonMap("text", createTextKeywordMultiField("text", MapperBuilderContext.ROOT))
+        );
+        RootObjectMapper mergeWith = createRootObjectMapper(
+            "_doc",
+            true,
+            Collections.singletonMap("text", createTextKeywordMultiField("text", MapperBuilderContext.ROOT))
+        );
+
+        final ObjectMapper merged = mergeInto.merge(mergeWith, MapperBuilderContext.ROOT);
+
+        TextFieldMapper text = (TextFieldMapper) merged.getMapper("text");
+        assertEquals("text", text.name());
+        assertEquals("text", text.simpleName());
+        KeywordFieldMapper keyword = (KeywordFieldMapper) text.multiFields().iterator().next();
+        assertEquals("text.keyword", keyword.name());
+        assertEquals("keyword", keyword.simpleName());
+    }
+
+    public void testMergeMultiFieldsWithinSubobjectsFalse() {
+        RootObjectMapper mergeInto = createRootObjectMapper(
+            "_doc",
+            true,
+            Collections.singletonMap("foo", createObjectSubobjectsFalseLeafWithMultiField())
+        );
+        RootObjectMapper mergeWith = createRootObjectMapper(
+            "_doc",
+            true,
+            Collections.singletonMap("foo", createObjectSubobjectsFalseLeafWithMultiField())
+        );
+
+        final ObjectMapper merged = mergeInto.merge(mergeWith, MapperBuilderContext.ROOT);
+
+        ObjectMapper foo = (ObjectMapper) merged.getMapper("foo");
+        ObjectMapper metrics = (ObjectMapper) foo.getMapper("metrics");
+        final TextFieldMapper textFieldMapper = (TextFieldMapper) metrics.getMapper("host.name");
+        assertEquals("foo.metrics.host.name", textFieldMapper.name());
+        assertEquals("host.name", textFieldMapper.simpleName());
+        FieldMapper fieldMapper = textFieldMapper.multiFields.iterator().next();
+        assertEquals("foo.metrics.host.name.keyword", fieldMapper.name());
+        assertEquals("keyword", fieldMapper.simpleName());
```

Comment:
I am not too sure that these additional merge tests belong to this class. I don't know whether there's an existing place for specific merge tests. Also, these are weird merge tests because they don't really test merging different mapper but rather merge the same mappers and verify the resulting simpleName and name. I can see how that will look weird in a couple of years: "why did we write these tests again?". Thoughts?

Reply:
Maybe add a comment to the effect that we're testing that leaf fields with dots in their names merge correctly?

Reply:
Really I think we should rewrite all these tests so that they extend MapperServiceTestCase and test the actual merge paths that MapperService calls, but let's do that as a followup.

Reply:
interesting, do they now test a different codepath? I mean they call root.merge directly, instead of going through mapper service, is that what you mean?